src/dataflow/src/source/postgres.rs

```diff
@@ -409,10 +458,67 @@ impl PostgresSourceReader {
                             }
                             self.metrics.lsn.set(self.lsn.into());
                         }
-                        Origin(_) | Relation(_) | Type(_) => {
+                        Relation(relation) => {
+                            let rel_id = relation.rel_id();
+                            if let Some(source_table) = self.source_tables.get(&rel_id) {
+                                // Start with the cheapest check first, this will catch the majority of alters
+                                if source_table.columns.len() != relation.columns().len() {
+                                    error!(
+                                        "alter table detected on {} with id {}",
+                                        source_table.name, source_table.relation_id
+                                    );
+                                    return Err(Fatal(anyhow!(
+                                        "source table {} with oid {} has been altered",
+                                        source_table.name,
+                                        source_table.relation_id
+                                    )));
+                                }
+                                let cols = relation.columns().iter().map(|c| PostgresColumn {
+                                    name: c.name().unwrap().to_string(),
+                                    type_oid: c.type_id(),
+                                    type_mod: c.type_modifier(),
+                                    nullable: bool::default(),
+                                    primary_key: bool::default(),
+                                });
+                                // Relation messages do not include nullability/primary_key data
+                                // so we check the name, type_oid, and type_mod explicitly and error if any of them differ
+                                if !source_table.columns.iter().zip(cols).all(|(src, rel)| {
+                                    src.name == rel.name
+                                        && src.type_oid == rel.type_oid
+                                        && src.type_mod == rel.type_mod
```

Comment:
Shouldn't we also check the nullable bit and the primary key bit? If yes it can also be written as `source_table.columns.iter().eq(cols)`

Reply:
The `Relation` message does not contain those pieces of data so we can't check them. I can reword the comment here to make this clearer though.

Reply:
Oh! I see what's going on now. I missed the comment. In that case the conversion to `PostgresColumn` can be avoided altogether:

```rust
if !source_table.columns.iter().eq_by(source_table.columns(), |src, rel| {
    src.name == rel.name().unwrap()
    && src.type_oid == rel.type_id()
    && src.type_mod == rel.type_modifier()
})
```

drops the `.to_string()` too